ci-operator/step-registry/cucushift/upgrade/check/capability/cucushift-upgrade-check-capability-commands.sh

```diff
@@ -0,0 +1,422 @@
+        return 1
+    fi
+
+    source_cap_files=$(grep -rl "capability.openshift.io/name" ${source_dir})
+    for source_file in ${source_cap_files}; do
+        filename=$(basename "${source_file}")
+        target_file="$target_dir/$filename"
+
+        if [ -f "$target_file" ]; then
+            source_cap=$(grep 'capability.openshift.io/name:' "$source_file" | cut -d ':' -f2 | tr -d '"' | sort | uniq )
+            target_cap=$(grep 'capability.openshift.io/name:' "$target_file" | cut -d ':' -f2 | tr -d '"' | sort | uniq )
+
+            if [ "$source_cap" != "$target_cap" ]; then
+                echo "Migrated capability in $filename: $source_cap -> $target_cap"
+                migrated_caps["$source_cap"]=$target_cap
+            fi
+        fi
+    done
+}
+
+function check_cvo_cap() {
+
+    local result=0
+    local capability_set=$1
+    local expected_status=$2
+    local cvo_field=$3
+
+    cvo_caps=$(oc get clusterversion version -o json | jq -rc "${cvo_field}")
+    if [[ "${capability_set}" == "" ]] && [[ "${cvo_caps}" != "null" ]]; then
+        echo "ERROR: ${expected_status} capability set is empty, but find capabilities ${cvo_caps} in cvo ${cvo_field}"
+        result=1
+    fi
+    if [[ "${capability_set}" != "" ]]; then
+        if [[ "${cvo_caps}" == "null" ]]; then
+            echo "ERROR: ${expected_status} capability set are ${capability_set}, but it's empty in cvo ${cvo_field}"
+            result=1
+        else
+            cvo_caps_str=$(echo $cvo_caps | tr -d '["]' | tr "," " " | xargs -n1 | sort -u | xargs)
+            if [[ "${cvo_caps_str}" == "${capability_set}" ]]; then
+                echo "INFO: ${expected_status} capabilities matches with cvo ${cvo_field}!"
+                echo -e "cvo_caps: ${cvo_caps_str}\n${expected_status} capability set: ${capability_set}"
+            else
+                echo "ERROR: ${expected_status} capabilities does not match with cvo ${cvo_field}!"
+                echo -e "cvo_caps: ${cvo_caps_str}\n${expected_status} capability set: ${capability_set}"
+                echo "diff [cvo_caps] [${expected_status} capability set]"
+                diff <( echo $cvo_caps_str | tr " " "\n" | sort | uniq) <( echo $capability_set | tr " " "\n" | sort | uniq )
+                result=1
+            fi
+        fi
+    fi
+
+    return $result
+} >&2
+
+if [ -f "${SHARED_DIR}/kubeconfig" ] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+else
+    echo "Unable to find kubeconfig under ${SHARED_DIR}!"
+    exit 1
+fi
+
+# Setup proxy if it's present in the shared dir
+if [[ -f "${SHARED_DIR}/proxy-conf.sh" ]]; then
+    # shellcheck disable=SC1091
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+# shellcheck disable=SC2207
+version_set=($(oc get clusterversion version -ojson | jq -r .status.history[].version))
+source_major_version=$( echo "${version_set[-1]}" | cut -f1 -d. )
```

Comment:
This is a separate step which is out of chain upgrade. So it's expected to be run once in chain upgrade. It validates the caps on the upgraded cluster compared to the initial cluster.

Reply:
do u mean, if it's 4.11-4.14 upgrade, it only do post check on cap for the last hop 4.14, while to compare with a 4.11 cluster? hmm, i wonder if the post check fail, is it easy for us to identify which hop got a wrong cap.

Reply:
Yeah, it's the drawback. But I don't prefer to move it to upgrade step as the step is complicated enough and it will be more complex with more caps introduced. It's not cvo only feature and all cap related components are supposed to ensure the functionality. I'm hoping other qe teams to treat it carefully in their e2e automation.

Reply:
do we have a proper job for a test against chain upgrade on such profile?

Reply:
We don't have chain upgrade job on caps profile currently. If someday we decide to add such job, we can update the step as needed, but I'm afraid that would be complex